### PR TITLE
Validate distribution base paths

### PR DESCRIPTION
**Pull request by daviddavis**
fixes #3412
https://pulp.plan.io/issues/3412

### CODE REVIEW

File: pulpcore/pulpcore/app/serializers/repository.py [hunk 1]
```diff
@@ -249,6 +249,9 @@ class Meta:
             'name', 'base_path', 'publisher', 'publication', 'base_url', 'repository',
```
Comment (dralley) [inline pulpcore/pulpcore/app/serializers/repository.py]: It's a bit strange that we have ```base_path``` mapped to ```base_url``` instead of using one name consistently (not part of this PR specifically)

File: pulpcore/pulpcore/app/serializers/repository.py [hunk 2]
```diff
@@ -249,6 +249,9 @@ class Meta:
...
         )
 
+    def validate_base_path(self, path):
+        return self._validate_relative_path(path)
+
```
Comment (dralley) [inline pulpcore/pulpcore/app/serializers/repository.py]: Why do this separately from the ```validate()``` we use for everything else?

File: pulpcore/pulpcore/app/serializers/repository.py (hunk 1 above)
Comment (daviddavis) [reply pulpcore/pulpcore/app/serializers/repository.py]: I'm not sure I understand. They're two separate things: base_path is like `foo/bar` and base_url is the full url like `localhost:8000/foo/bar`.

File: pulpcore/pulpcore/app/serializers/repository.py (hunk 2 above)
Comment (daviddavis) [reply pulpcore/pulpcore/app/serializers/repository.py]: Yes, DRF autodetects these methods:

http://www.django-rest-framework.org/api-guide/serializers/#field-level-validation

The validate method is not field specific. In this case, it's checking two fields: publisher and repository. It's better to use field-specific validation as it sets the error on the particular failing field in the response. Also, you're suppose to raise an exception which halts execution so adding this to validate would run the other validation on publisher and repository.

File: pulpcore/pulpcore/app/serializers/repository.py (hunk 1 above)
Comment (dralley) [reply pulpcore/pulpcore/app/serializers/repository.py]: I was referring to:

```
 base_url = BaseURLField(
        source='base_path', read_only=True,
        help_text=_('The URL for accessing the publication as defined by this distribution.')
)
```

But I think maybe I'm just not understanding the code semantics.

File: pulpcore/pulpcore/app/serializers/repository.py (hunk 2 above)
Comment (dralley) [reply pulpcore/pulpcore/app/serializers/repository.py]: Ah, I was looking at the *Django* docs, not DRF.  My bad.  And that makes sense.

**Comment by Ichimonji10** [comment]
Does this PR mean that if I send an invalid base_path to Pulp with a trailing slash, that Pulp will silently change what I've send and save a base path that doesn't have a trailing slash? If so, I dislike this solution. When I send a request to Pulp, I expect that request to be interpreted as is. Magical changes being made without my knowledge are extremely frustrating.

In other words, loudly complain when my code does something wrong, so that I can fix it. Don't assume that I *really* meant for my code to do something else.

**Comment by daviddavis** [comment]
The problem I ran into is that 'my-base-path/' is actually a valid base path:

```
In [1]: from django.core.validators import URLValidator
In [2]: validate = URLValidator()
In [3]: validate("http://localhost/content/my-base-path//test.iso")
(no exception)
```

I could add some validation specifically for Pulp (which I'm a bit apprehensive to do) or fix the distribution code so that it properly handles requests to `http://localhost/content/my-base-path//test.iso`. I'm a bit apprehensive about the latter as well though as I think most users will assume they need to add a slash since they are specifying a path (and paths end in slashes).

Thoughts @dralley @Ichimonji10 ?

**Comment by Ichimonji10** [comment]
I think letting base paths have trailing slashes - or even requiring that they have slashes - would lead to a more usable application. Consider these code snippets:

```python
from urllib.parse import urljoin

# In a real client-side application, `distribution` comes from Pulp's API.
distribution = {'base_url': 'pulp-3.example.com:8000/content/my-base-path'}
base_url = 'http://' + distribution['base_url']
basename = '1.iso'

print(base_url + basename)
print(urljoin(base_url, basename))
```

The results are not as intended:

```
http://pulp-3.example.com:8000/content/my-base-path1.iso
http://pulp-3.example.com:8000/content/1.iso
```

On the other hand, if base paths have trailing slashes, results are sane:

```python
from urllib.parse import urljoin

# In a real client-side application, `distribution` comes from Pulp's API.
distribution = {'base_url': 'pulp-3.example.com:8000/content/my-base-path/'}
base_url = 'http://' + distribution['base_url']
basename = '1.iso'

print(base_url + basename)
print(urljoin(base_url, basename))
```

Output:

```
http://pulp-3.example.com:8000/content/my-base-path/1.iso
http://pulp-3.example.com:8000/content/my-base-path/1.iso
```

I discuss this in [Pulp #3412](https://pulp.plan.io/issues/3412)`.

**Comment by daviddavis** [comment]
Relying on the precise science of [internet polling](https://etherpad.net/p/D2UjRxz01J), I'm going to update this PR to check for slashes at the beginning/end of paths.

**Comment by Ichimonji10** [comment]
+0

I still think there's a usability issue here. But at least we're telling users how they should use using Pulp.

File: pulpcore/pulpcore/app/serializers/base.py [hunk 3]
```diff
@@ -150,6 +154,31 @@ def _populate_generic_fields(self, instance, field_mappings):
...
 
+    def _validate_relative_path(self, path):
+        """
+        Validate a relative path (eg from a url) and strip out beginning/end slashes
+
+        Args:
+            path (str): A relative path to validate
+
+        Returns:
+            str: the validated path stripped of beginning and end slashes
```
Comment (Ichimonji10) [inline pulpcore/pulpcore/app/serializers/base.py]: This needs to be updated, correct?
